workflows/labels: add bottleneck for throttling

wolfgangwalther opened the pull request:
This tries to improve the situation around API limits, which hit us in #417463.

In the octokit repo, I found this https://github.com/octokit/octokit.js/issues/1069#throttling, so I added `bottleneck`.

## Things done

- [x] Tested this in my fork.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/labels: log current API rate_limits

This will give us a little bit better insight into how close we're
running to the limits.

wolfgangwalther committed:
workflows/labels: add bottleneck for throttling

This should prevent us from hitting API rate limits, even when we run
huge manual jobs. It just takes a bit longer.

philiptaron reviewed (commented):
This is a really interesting approach. I really like how easy it was to add, and if it works, points the way forward.

wolfgangwalther committed:
workflows/labels: improve logging for skipped PRs

Conditions that cause a PR to be skipped are now marked clearly in the
log output.

MattSturgeon commented on .github/workflows/labels.yml:
```diff
@@ -39,20 +39,43 @@ jobs:
+              reservoirRefreshInterval: 60 * 60 * 1000
+            })
+            // Pause between mutative requests
+            const writeLimits = new Bottleneck({ minTime: 1000 }).chain(allLimits)
```
`.chain(allLimits)` is _fancy_

MattSturgeon commented on .github/workflows/labels.yml:
```diff
@@ -39,20 +39,43 @@ jobs:
+            const allLimits = new Bottleneck({
+              // Avoid concurrent requests
+              maxConcurrent: 1,
+              reservoir: 2500, // Half of the actual limit, because other jobs need some, too!
```
Can the comment change to "hourly" limit instead of "actual" limit, just to be clearer?

We could even add a link to the relevant API limit docs :rocket: 

> We should go lower, like 1500 imo

The last run was ~25 PRs. If we assume an average of 30, how many requests would that be?

If the numbers are reasonable, we could calibrate the reservoir such that a typical run would not be affected by our internal bottleneck, while still being maximally restrictive to avoid atypical runs impacting on our overall limits.

philiptaron commented:
> One thing that is not taken into account is the artifact download, because it happens via the separate `artifactClient` and thus doesn't use the limiter. I'll look into that again, but I think it's not possible to share the octokit instance for both :/.

Yet another reason that suggests we ought to use an approach that doesn't need the artifact download... though I suppose downloading the diff is an equivalent number of requests.

MattSturgeon reviewed (commented):
Definitely a better approach than I proposed (https://github.com/NixOS/nixpkgs/pull/417463#issuecomment-2979931843 and https://github.com/NixOS/nixpkgs/pull/417463#issuecomment-2980002001)!

> One thing that is not taken into account is the artifact download, because it happens via the separate `artifactClient` and thus doesn't use the limiter. I'll look into that again, but I think it's not possible to share the octokit instance for both :/.

I think having a conservative bottleneck _should_ help with this, although there will be edge-cases I've not considered.

MattSturgeon commented:
> the main reason we're using it is, because I sure don't want to deal with downloading and extracting a zip-file

I agree this is best avoided. I'm surprised there isn't a decent API built into node, but there are a lot of libraries on the npm as always. [adm-zip](https://www.npmjs.com/package/adm-zip) looks like a simple option with no transitive dependencies. So options available if we _do_ need to go down that route.

wolfgangwalther committed:
workflows/labels: add simple stats

To ease debugging rate-limiting issues.

philiptaron reviewed (approved):
If the test goes successfully, let's see how it does in production.

MattSturgeon commented on .github/workflows/labels.yml:
```diff
@@ -101,10 +134,14 @@ jobs:
               async (response, done) => (await Promise.allSettled(response.data.map(async (pull_request) => {
                 try {
-                  const log = (k,v) => core.info(`PR #${pull_request.number} - ${k}: ${v}`)
+                  const log = (k,v,skip) => {
+                    core.info(`PR #${pull_request.number} - ${k}: ${v}` + skip ? ' (skipped)' : '')
+                    return skip
+                  }
```
Interesting that all uses of `log` make sense to accept a `skip` parameter... Feels like a strange abstraction, but it works here :+1:

MattSturgeon commented on .github/workflows/labels.yml:
```diff
@@ -39,20 +39,53 @@ jobs:
+
+      - name: Log current API rate limits
+        env:
+          GH_TOKEN: ${{ github.token }}
```
Good catch!

MattSturgeon commented:
> The unix timestamp is `Tue Jun 17 2025 14:45:14 GMT+0000`. So roughly 15 minutes in

Looks like the job ran at 13:57 and the reset is at 14:45, so yes we're roughly 15m into our hourly window... And we've already made a fair number of requests.

If we extrapolate out to an hour, `650*4` is 2600, so roughly half our ratelimit. I can definitely see this creeping up over time, if we're not careful; both due to increasing workflows and increasing PR events.

MattSturgeon commented:
I know in some scenarios the graphql API can be more efficient, because you specify exactly what data you wish to recieve. I believe that's why it has a [separate](https://docs.github.com/en/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api) ratelimit.

It can be a pain to figure out what the query should look like just from reading the [reference docs](https://docs.github.com/en/graphql/reference/queries), but at least the [explorer](https://docs.github.com/en/graphql/overview/explorer) has some sane auto-complete.

Another thing worth considering is whether GitHub Apps get any ratelimit separation. I believe they don't, since they count as being part of whatever "installation" they are installed and running on.

The Search API also has a [separate](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#rate-limit) (albeit lower) ratelimit,[^1] and can be used either via [REST](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-issues-and-pull-requests) or [graphql](https://docs.github.com/en/graphql/reference/queries#search).

I was actually wondering if using a search query would work for https://github.com/NixOS/nixpkgs/pull/417512#discussion_r2152249375, since you can specify [`updated`](https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests#search-by-when-an-issue-or-pull-request-was-created-or-last-updated) with a `>=`, E.g. `"type:pr AND state:open AND updated:>=2024-06-17T16:01:00"`

[^1]: At least the REST API does... The graphql search API may be treated as a normal graphql query; I'm unsure.

MattSturgeon commented:
If using the Search API, we should be aware of this [recent changelog](https://github.blog/changelog/2025-03-06-github-issues-projects-api-support-for-issues-advanced-search-and-more/#clock1-timeline-events-for-issue-types).

TL;DR issue/pr search is transitioning to "advanced" search syntax (IIUC this means support for `"AND"`, `"OR"` keywords), right now this is `advanced_search=true` in REST requests and `type: ISSUE_ADVANCED` in graphql.

But as of the 4th of September, these will be removed and advanced search becomes the default behaviour. IIUC there is no transition period where the default behaviour changes _before_ the explicit parameters get deprecated and removed...
